### PR TITLE
[sweet][android] Implement `groupView`

### PR DESCRIPTION
**Pull request by barthap**
# Why

Resolves ENG-4122

We need to support some `ViewGroupManager` specific methods.

# How

Introduced a `viewGroup` definition inside the `viewManager` definition (example below contains update from c169f47):

```kt
class MyModule : Module {
  override fun definition() = ModuleDefinition {
    // ...

    viewManager {
      // ...
      vievGroup {
        addChildView { parent, child, index -> ... }
        getChildCount { view -> ... }
        getChildViewAt { parent, index -> ... }
        removeChildView { parent, view -> ... }
        removeChildViewAt { parent, index -> ... }
      }
    }
  }
}
```

When the managed `View` is not an instance of `ViewGroup`, the whole section is ignored.

# Test Plan

We have no expo-module supporting this yet (expo-maps will be first) so now I faked expo-linear-gradient to be a view group to test if module methods are called.

### CODE REVIEW

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/views/ViewManagerDefinitionBuilder.kt [hunk 1]
```diff
@@ -65,4 +67,15 @@ class ViewManagerDefinitionBuilder {
   fun events(vararg callbacks: String) {
     callbacksDefinition = CallbacksDefinition(callbacks)
   }
+
+  /**
+   * Creates the group view definition that scopes group view-related definitions.
+   */
+  inline fun groupView(body: GroupViewDefinitionBuilder.() -> Unit) {
```
Comment (barthap) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/views/ViewManagerDefinitionBuilder.kt:74]: `groupView` or maybe `viewGroup`?

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/views/GroupViewDefinitionBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import expo.modules.kotlin.modules.DefinitionMarker
+
+@DefinitionMarker
+class GroupViewDefinitionBuilder {
+  @PublishedApi
+  internal var actions = mutableMapOf<GroupViewAction.Action, GroupViewAction>()
+
+  fun build() = GroupViewDefinition(actions)
+
+  inline fun <reified ParentViewType : ViewGroup, reified ChildViewType : View> addView(
```
Comment (barthap) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/views/GroupViewDefinitionBuilder.kt:16]: maybe `addChildView`?

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/views/GroupViewAction.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  )
+
+  enum class Action {
+    ADD_VIEW,
+    GET_CHILD_AT,
+    GET_CHILD_COUNT,
+    REMOVE_VIEW,
+    REMOVE_VIEW_AT
+  }
+}
```
Comment (barthap) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/views/GroupViewAction.kt:22]: no idea for namings here 😞

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/views/GroupViewAction.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+class GroupViewAction(
+  val body: (Payload) -> Any
+) {
+  data class Payload(
+    val parentView: ViewGroup? = null,
+    val childView: View? = null,
+    val index: Int? = null
+  )
+
+  enum class Action {
```
Comment (barthap) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/views/GroupViewAction.kt:15]:
```suggestion
  enum class Kind {
```
🤔

**Review by lukmccall — COMMENTED** [review]
It looks good. However, I'm not sure if a more straightforward approach won't be better here. I'm referring to the way how all the methods connected with the group view are passed to the wrapper delegate. I don't know if I like the generic `callGroupViewActionOrElse` method 🤔

**Comment by barthap** [comment]
> I'm not sure if a more straightforward approach won't be better here. I'm referring to the way how all the methods connected with the group view are passed to the wrapper delegate.

Could you elaborate? Do you think it's better to keep `GroupViewDefinition` having just a few closure references as separate variables instead of having them stored in a map?

**Comment by lukmccall** [comment]
> Do you think it's better to keep GroupViewDefinition having just a few closure references as separate variables instead of having them stored in a map?

Yes, exactly. I think it will be easier to track. Also, the error stack trace will look better without the generic dispatcher. What do you think?

**Comment by barthap** [comment]
As there are only five of them and their amount is not that likely to change in the nearest future, I agree - such generic solution is not necessary here. I'll simplify it then.

**Review by lukmccall — APPROVED** [review]

**Pull request merged**